shadowsocks/client/packet_listener.go

```diff
@@ -0,0 +1,130 @@
+var udpPool = slicepool.MakePool(clientUDPBufferSize)
+
+type packetListener struct {
+	endpoint onet.PacketEndpoint
+	cipher   *ss.Cipher
+}
+
+func NewShadowsocksPacketListener(endpoint onet.PacketEndpoint, cipher *ss.Cipher) (onet.PacketListener, error) {
+	if endpoint == nil {
+		return nil, errors.New("Argument endpoint must not be nil")
+	}
+	if cipher == nil {
+		return nil, errors.New("Argument cipher must not be nil")
+	}
+	return &packetListener{endpoint: endpoint, cipher: cipher}, nil
+}
+
+func (c *packetListener) ListenPacket(ctx context.Context) (net.PacketConn, error) {
+	proxyConn, err := c.endpoint.Connect(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("Could not connect to endpoint: %x", err)
+	}
+	conn := packetConn{Conn: proxyConn, cipher: c.cipher}
+	return &conn, nil
+}
+
+type packetConn struct {
+	net.Conn
+	cipher *ss.Cipher
+}
+
+// WriteTo encrypts `b` and writes to `addr` through the proxy.
+func (c *packetConn) WriteTo(b []byte, addr net.Addr) (int, error) {
+	socksTargetAddr := socks.ParseAddr(addr.String())
+	if socksTargetAddr == nil {
+		return 0, errors.New("Failed to parse target address")
+	}
+	lazySlice := udpPool.LazySlice()
+	cipherBuf := lazySlice.Acquire()
+	defer lazySlice.Release()
+	saltSize := c.cipher.SaltSize()
+	// Copy the SOCKS target address and payload, reserving space for the generated salt to avoid
+	// partially overlapping the plaintext and cipher slices since `Pack` skips the salt when calling
+	// `AEAD.Seal` (see https://golang.org/pkg/crypto/cipher/#AEAD).
+	plaintextBuf := append(append(cipherBuf[saltSize:saltSize], socksTargetAddr...), b...)
+	buf, err := ss.Pack(cipherBuf, plaintextBuf, c.cipher)
+	if err != nil {
+		return 0, err
+	}
+	_, err = c.Conn.Write(buf)
+	return len(b), err
+}
+
+// ReadFrom reads from the embedded PacketConn and decrypts into `b`.
+func (c *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
+	lazySlice := udpPool.LazySlice()
+	cipherBuf := lazySlice.Acquire()
+	defer lazySlice.Release()
+	n, err := c.Conn.Read(cipherBuf)
+	if err != nil {
+		return 0, nil, err
+	}
+	// Decrypt in-place.
+	buf, err := ss.Unpack(nil, cipherBuf[:n], c.cipher)
+	if err != nil {
+		return 0, nil, err
+	}
+	socksSrcAddr := socks.SplitAddr(buf)
+	if socksSrcAddr == nil {
+		return 0, nil, errors.New("Failed to read source address")
+	}
+	srcAddr := newAddr(socksSrcAddr.String(), "udp")
+	n = copy(b, buf[len(socksSrcAddr):]) // Strip the SOCKS source address
+	if len(b) < len(buf)-len(socksSrcAddr) {
+		return n, srcAddr, io.ErrShortBuffer
+	}
+	return n, srcAddr, nil
+}
+
+type addr struct {
+	address string
+	network string
+}
+
+func (a *addr) String() string {
+	return a.address
+}
+
+func (a *addr) Network() string {
+	return a.network
+}
+
+// newAddr returns a net.Addr that holds an address of the form `host:port` with a domain name or IP as host.
+// Used for SOCKS addressing.
+func newAddr(address, network string) net.Addr {
```

Comment:
*nit* just wondering can we reuse `net.UDPAddr` and `net.TCPAddr` in the standard library?

Reply:
Maybe. In theory the Shadowsocks server could return a domain name as the host address. In practice that won't happen if you don't use a domain name as the address.

Perhaps we need separate APIs for dialing with a host versus dialing with an IP. For APIs that require IP-Port, there's https://pkg.go.dev/net/netip@go1.20.2#AddrPort.

Humm... That requires more thought. Let's ponder on this, I won't submit right now.

Reply:
Right and I think we can always parse the hostname to IP:port.

Reply:
To close this, I think we should go ahead with allowing host names and IPs as a string. All PacketListeners/StreamDialer must support both. If it doesn't, it should be a different interface that only takes the IP. In that case, users will need to adapt by either rejecting the request with an invalid address error, or resolving the host name.